core/shared/src/main/scala/spire/algebra/Gcd.scala

```diff
@@ -0,0 +1,25 @@
+package spire.algebra
+
+import annotation.tailrec
+import scala.{specialized => sp}
+
+import spire.syntax.eq._
+import spire.syntax.euclideanRing._
+
+trait Gcd[@sp(Int, Long, Float, Double) A] extends Any {
+  def gcd(a: A, b: A): A
+  def lcm(a: A, b: A): A
+
+  @tailrec final def euclid(a: A, b: A)(implicit eq: Eq[A], er: EuclideanRing[A]): A =
+    if (b === er.zero) a else euclid(b, a % b)
+}
+
+object Gcd {
+  @inline final def apply[A](implicit ev: Gcd[A]): Gcd[A] = ev
+
+  implicit def fromEuclideanRing[@sp(Int, Long, Float, Double) A: EuclideanRing: Eq] =
```

Comment:
Let's not make this `implicit`. I'd much rather people explicitly opt-in to this. For example, rational numbers require quite a bit of care, and polynomials need to deal with an edge case or 2.